tcmalloc/internal/percpu_tcmalloc.h

```diff
@@ -500,13 +632,144 @@ static inline ABSL_ATTRIBUTE_ALWAYS_INLINE void* TcmallocSlab_Pop(
 }
 #endif  // defined(__x86_64__)
 
+#if defined(__aarch64__)
+template <size_t Shift, size_t NumClasses>
+static inline ABSL_ATTRIBUTE_ALWAYS_INLINE void* TcmallocSlab_Pop(
+    typename TcmallocSlab<Shift, NumClasses>::Slabs* slabs, size_t cl,
+    UnderflowHandler f) {
+  void* result;
+  void* region_start;
+  uint64_t cpu_id;
+  void* begin_ptr;
+  uintptr_t current;
+  uintptr_t new_current;
+  uintptr_t begin;
+  // Multiply cl by the bytesize of each header
+  size_t cl_lsl3 = cl * 8;
+#if TCMALLOC_PERCPU_USE_RSEQ_ASM_GOTO_OUTPUT
+  asm goto
+#else
+  bool underflow;
+  asm
+#endif
+      (
+          // TODO(b/141629158):  __rseq_cs only needs to be writeable to allow
+          // for relocations, but could be read-only for non-PIE builds.
+          ".pushsection __rseq_cs, \"aw?\"\n"
+          ".balign 32\n"
+          ".local __rseq_cs_TcmallocSlab_Pop_%=\n"
+          ".type __rseq_cs_TcmallocSlab_Pop_%=,@object\n"
+          ".size __rseq_cs_TcmallocSlab_Pop_%=,32\n"
+          "__rseq_cs_TcmallocSlab_Pop_%=:\n"
+          ".long 0x0\n"
+          ".long 0x0\n"
+          ".quad 4f\n"
+          ".quad 5f - 4f\n"
+          ".quad 2f\n"
+          ".popsection\n"
+#if !defined(__clang_major__) || __clang_major__ >= 9
+          ".reloc 0, R_AARCH64_NONE, 1f\n"
+#endif
+          ".pushsection __rseq_cs_ptr_array, \"aw?\"\n"
+          "1:\n"
+          ".balign 8;"
+          ".quad __rseq_cs_TcmallocSlab_Pop_%=\n"
+          // Force this section to be retained.  It is for debugging, but is
+          // otherwise not referenced.
+          ".popsection\n"
+          ".pushsection .text.unlikely, \"ax?\"\n"
+          ".long %c[rseq_sig]\n"
+          ".local TcmallocSlab_Pop_trampoline_%=\n"
+          ".type TcmallocSlab_Pop_trampoline_%=,@function\n"
+          "TcmallocSlab_Pop_trampoline_%=:\n"
+          "2:\n"
+          "b 3f\n"
+          ".popsection\n"
+          // Prepare
+          "3:\n"
+          // Use current as scratch here to hold address of this function's
+          // critical section
+          "adrp %[current], __rseq_cs_TcmallocSlab_Pop_%=\n"
+          "add  %[current], %[current], :lo12:__rseq_cs_TcmallocSlab_Pop_%=\n"
+          "str %[current], [%[rseq_abi], %c[rseq_cs_offset]]\n"
+          // Start
+          "4:\n"
+          // cpu_id = __rseq_abi.cpu_id;
+          "ldr %w[cpu_id], [%[rseq_abi], %[rseq_cpu_offset]]\n"
+          // region_start = Start of cpu region
+          "add %[region_start], %[slabs], %[cpu_id], LSL %c[shift]\n"
+          // begin_ptr = &(slab_headers[0]->begin)
+          "add %[begin_ptr], %[region_start], #4\n"
+          // current = slab_headers[cl]->current (current index)
+          "ldrh %w[current], [%[region_start], %[cl_lsl3]]\n"
+          // begin = slab_headers[cl]->begin (begin index)
+          "ldrh %w[begin], [%[begin_ptr], %[cl_lsl3]]\n"
+          "sub %w[new_current], %w[current], #1\n"
```

Comment:
Should this be sunk below the branch?

Reply:
(Or between the comparison and the branch to hide latency?)

Reply:
I will try between the compare and the branch. I had it below the branch originally, but this seemed to be better.

Reply:
I tried a couple of different approaches, very little differences but I think overall putting it in between the cmp and branch seem to offer the very very slightly better performance.